client/sessioniface/fetcher.go

```diff
@@ -0,0 +1,18 @@
+package sessioniface
+
+import (
+	"context"
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	exchange "github.com/ipfs/go-ipfs-exchange-interface"
+)
+
+type AddRemoveCid interface {
+	IsAdd() bool
+	Key() cid.Cid
+}
+
+type ChannelFetcher interface {
+	exchange.Fetcher
+	GetBlocksCh(ctx context.Context, keys <-chan AddRemoveCid) (<-chan blocks.Block, error)
```

Comment:
I wonder if this would be more ergonomic to work against with separate add and remove channels of cid.Cid's directly?

Reply:
This seems like a bad idea since if a CID is added and removed at a similar time they could come out of order which would result in a state the caller did not intend (e.g. a CID being requested they thought they'd removed, or a CID being removed they'd thought had been requested).

You could hide the calls behind a single object with multiple functions (option 2 here https://github.com/ipfs/go-bitswap/pull/593#issue-1437742720) since mutexes could guard you. It does come with some complexity and extra locking though

Reply:
good point! it's a shame golang's generics don't make it easier to specify at type around `AddRemoveCid` and we're stuck with accessor functions